### PR TITLE
fix the lpr linting problem

### LINKED ISSUE #54: in lpr example lint passes in CI but not when I build locally
The CI passes no problem. But if I build the cookbook locally, the linter is failing.

```
(base) davidetisi@MacBook-Pro-di-Davide-2:~/Documents/Work/Software/software-cookbook$ nox -e lint
nox > Running session lint
nox > Re-using existing virtual environment at .nox/lint.
nox > black --check --diff ipynb-to-gallery.py generate-gallery.py noxfile.py docs/src/conf.py examples
Skipping .ipynb files as Jupyter dependencies are not installed.
You can fix this by running ``pip install "black[jupyter]"``
--- /Users/davidetisi/Documents/Work/Software/software-cookbook/examples/lpr/datasets/runner.py	2023-06-19 15:05:32+00:00
+++ /Users/davidetisi/Documents/Work/Software/software-cookbook/examples/lpr/datasets/runner.py	2024-03-29 09:43:45.592012+00:00
@@ -9,21 +9,22 @@
 from ase.calculators.singlepoint import SinglePointCalculator
 from ase.parallel import paropen
 from ase.utils import basestring
 from ase import units
 
+
 def read_runner(fileobj, index=-1):
     """
     Read from a file in RuNNer format
 
     index is the frame to read, default is last frame (index=-1).
     """
     if isinstance(fileobj, str):
         fileobj = open(fileobj)
 
     if not isinstance(index, int) and not isinstance(index, slice):
-        raise TypeError('Index argument is neither slice nor integer!')
+        raise TypeError("Index argument is neither slice nor integer!")
 
     # If possible, build a partial index up to the last frame required
     last_frame = None
     if isinstance(index, int) and index >= 0:
         last_frame = index
@@ -35,17 +36,17 @@
     fileobj.seek(0)
     frames = []
     natoms = 0
     while fileobj:
         line = fileobj.readline()
-        if line.strip() == '':
+        if line.strip() == "":
             break
-        elif line.strip() == 'begin':
+        elif line.strip() == "begin":
             frame_pos = fileobj.tell()
-        elif line.split()[0] == 'atom':
+        elif line.split()[0] == "atom":
             natoms += 1
-        elif line.strip() == 'end':
+        elif line.strip() == "end":
             frames.append((frame_pos, natoms))
             natoms = 0
             if last_frame is not None and len(frames) > last_frame:
                 break
 
@@ -80,24 +81,28 @@
     for index in trbl:
         frame_pos, natoms = frames[index]
         fileobj.seek(frame_pos)
 
         # comment line
-        line = fileobj.readline() 
+        line = fileobj.readline()
         # Is there any valuable info to strip from the comment line?
         # info = key_val_str_to_dict(line)
 
         info = {}
         arrays = {}
 
-        cell = [] # Now we have to read 3 lines to extract information about the lattice
+        cell = (
+            []
+        )  # Now we have to read 3 lines to extract information about the lattice
 
         for ln in range(3):
             line = fileobj.readline()
             vals = line.split()
             cell.append(vals[1:4])
-        cell = np.array(cell,dtype='float')*units.Bohr # This converts to Angstrom, which is the default unit in ASE
+        cell = (
+            np.array(cell, dtype="float") * units.Bohr
+        )  # This converts to Angstrom, which is the default unit in ASE
 
         positions = []
         symbols = []
         forces = []
         numbers = []
@@ -110,86 +115,122 @@
 
             symbols.append(vals[4])
             positions.append(coords)
             forces.append(tmpforces)
 
-
         try:
-            positions = np.array(positions)*units.Bohr
-            forces = np.array(forces)*units.Hartree/units.Bohr # Transform from Ha/bohr to eV/angstrom
+            positions = np.array(positions) * units.Bohr
+            forces = (
+                np.array(forces) * units.Hartree / units.Bohr
+            )  # Transform from Ha/bohr to eV/angstrom
 
         except TypeError:
-            raise IOError('Badly formatted data, ' +
-                          'or end of file reached before end of frame')
-
-
-        line = fileobj.readline()
-        energy = float(line.split()[1])*units.Hartree
+            raise IOError(
+                "Badly formatted data, " + "or end of file reached before end of frame"
+            )
+
+        line = fileobj.readline()
+        energy = float(line.split()[1]) * units.Hartree
         line = fileobj.readline()
         charge = float(line.split()[1])
 
-        arrays['forces'] = forces
-        info['nrg'] = energy
+        arrays["forces"] = forces
+        info["nrg"] = energy
 
         structure = Atoms(symbols=symbols, positions=positions, cell=cell, pbc=True)
 
         calc = SinglePointCalculator(structure, energy=energy, forces=forces)
         structure.set_calculator(calc)
 
-
-        #atoms = Atoms(symbols=symbols,
+        # atoms = Atoms(symbols=symbols,
         #              positions=positions,
         #              cell=cell,
         #              pbc=pbc,
         #              info=info)
 
         yield structure
 
-def write_runner(fileobj,images,comment=None,append=False):
+
+def write_runner(fileobj, images, comment=None, append=False):
     """
     Write output in runner format. Written quickly with no regard to the form
     """
     if isinstance(fileobj, basestring):
-        mode = 'w'
+        mode = "w"
         if append:
-            mode = 'a'
-        fileobj = open(fileobj,mode)
-
-    if hasattr(images, 'get_positions'):
+            mode = "a"
+        fileobj = open(fileobj, mode)
+
+    if hasattr(images, "get_positions"):
         images = [images]
 
     for atoms in images:
 
-        try: forces = atoms.get_forces()/units.Hartree * units.Bohr
-        except: 
+        try:
+            forces = atoms.get_forces() / units.Hartree * units.Bohr
+        except:
             sys.stderr.write("No forces found, setting them to 0\n")
             nat = atoms.get_number_of_atoms()
-            forces = np.zeros((nat,3))
-
-        try: energy = atoms.get_potential_energy()/units.Hartree
+            forces = np.zeros((nat, 3))
+
+        try:
+            energy = atoms.get_potential_energy() / units.Hartree
         except:
             sys.stderr.write("No energy found, setting it to 0\n")
             energy = 0
 
-        #atoms.wrap()
-        fileobj.write('begin\n')
+        # atoms.wrap()
+        fileobj.write("begin\n")
         if comment is None:
-            fileobj.write('comment ' + str(atoms.get_number_of_atoms())  + ' atoms, species ' + str(atoms.get_chemical_formula()) + "\n")
+            fileobj.write(
+                "comment "
+                + str(atoms.get_number_of_atoms())
+                + " atoms, species "
+                + str(atoms.get_chemical_formula())
+                + "\n"
+            )
         else:
-            fileobj.write('comment ' + comment)
-
-        cell = atoms.get_cell()/units.Bohr
+            fileobj.write("comment " + comment)
+
+        cell = atoms.get_cell() / units.Bohr
         for idx in xrange(3):
-            fileobj.write("lattice " + str(cell[idx][0]) + " " + str(cell[idx][1]) + " " + str(cell[idx][2]) + "\n")
+            fileobj.write(
+                "lattice "
+                + str(cell[idx][0])
+                + " "
+                + str(cell[idx][1])
+                + " "
+                + str(cell[idx][2])
+                + "\n"
+            )
 
         nat = atoms.get_number_of_atoms()
 
-        positions = atoms.get_positions(wrap=True)/units.Bohr
+        positions = atoms.get_positions(wrap=True) / units.Bohr
 
         for idx in xrange(nat):
-            #fileobj.write('atom ' + atoms[idx].position + atoms[idx].symbol + '0.000000 0.000000 ' + forces[idx])
-            fileobj.write("atom " + str(positions[idx,0]) + " " + str(positions[idx,1]) + " " + str(positions[idx,2]) + " " + atoms[idx].symbol + " " + str(0.0000000) + " " + str(0.000000) + " " + str(forces[idx][0]) + " " + str(forces[idx][1]) + " " + str(forces[idx][2]) + "\n")
-
-
-        fileobj.write('energy ' + str(energy) + "\n")
-        fileobj.write('charge 0.0000\n')
-        fileobj.write('end\n')
+            # fileobj.write('atom ' + atoms[idx].position + atoms[idx].symbol + '0.000000 0.000000 ' + forces[idx])
+            fileobj.write(
+                "atom "
+                + str(positions[idx, 0])
+                + " "
+                + str(positions[idx, 1])
+                + " "
+                + str(positions[idx, 2])
+                + " "
+                + atoms[idx].symbol
+                + " "
+                + str(0.0000000)
+                + " "
+                + str(0.000000)
+                + " "
+                + str(forces[idx][0])
+                + " "
+                + str(forces[idx][1])
+                + " "
+                + str(forces[idx][2])
+                + "\n"
+            )
+
+        fileobj.write("energy " + str(energy) + "\n")
+        fileobj.write("charge 0.0000\n")
+        fileobj.write("end\n")
would reformat /Users/davidetisi/Documents/Work/Software/software-cookbook/examples/lpr/datasets/runner.py

Oh no! 💥 💔 💥
1 file would be reformatted, 10 files would be left unchanged.
nox > Command black --check --diff ipynb-to-gallery.py generate-gallery.py noxfile.py docs/src/conf.py examples failed with exit code 1
nox > Session lint failed.
```

### PR DESCRIPTION
**Pull request by DavideTisi**
This pr fixes #54. It is a hack; it generates the list of files you want to lint. maybe we can do better.
Of course, this code is 100% done by me and not at all generated by chatgpt.

### CODE REVIEW

**Review by PicoCentauri — COMMENTED** [review]
We should avoid blowing up our config file with too much logic. @Luthaf already raised the same concern.

Is there no way that flake ignores files that are given in the gitignore.

**Comment by SanggyuChong** [comment]
Guys, let me know if there's a much better way around all this hassle. Should we host these files altogether elsewhere?

**Review by SanggyuChong — COMMENTED** [review]
Thanks for adding the `.gitignore` for our example. Regarding all the other changes to the `noxfile.py`, would this be generally applicable to all the other examples as well then? If so, if this is a once-and-for-all solution, perhaps we do this now...

**Comment by DavideTisi** [comment]
> Is there no way that flake ignores files that are given in the gitignore.

this is what the code does basically. 
According to chatgpt the problem is not flake but also black etc

The reason I add to add the `if` was that if it was just reading from `.gitignore` then he started to linting XYZ files. 
Those are not skipped but still, I do not want to lint them

**Comment by DavideTisi** [comment]
> Thanks for adding the `.gitignore` for our example. Regarding all the other changes to the `noxfile.py`, would this be generally applicable to all the other examples as well then? If so, if this is a once-and-for-all solution, perhaps we do this now...

This works in general the only thing is when you add a file that you want to keep track that you do not want to lint you have to add to the exclusion `if clause`

File: noxfile.py [hunk 1]
```diff
@@ -30,6 +22,70 @@
...
+        folder,
+    ]
+    ignored_files_output = get_command_output(ignored_files_command, cwd=folder)
+
+    # Get the list of all tracked files
+    tracked_files_command = ["git", "ls-files", folder]
+    tracked_files_output = get_command_output(tracked_files_command, cwd=folder)
+
+    # Filter the tracked files to exclude ignored ones
+    filtered_files = filter_files(tracked_files_output, ignored_files_output)
```
Comment (Luthaf) [inline noxfile.py]: This logic is a bit weird to me. Doesn't `git ls-file` return a list **without** the ignored files already?

**Comment by Luthaf** [comment]
If there is a flag in flake8/black to make them respect `.gitignore`, I'd say we should use that. Otherwise, we'll need to put the config somewhere unfortunately =/

We could also have the examples download everything they need in `downloaded/`, and exclude this folder from both git and linting.

**Comment by SanggyuChong** [comment]
I like @Luthaf's second proposal. @DavideTisi, let me know if you want me to take care of this.

File: noxfile.py (hunk 1 above)
Comment (DavideTisi) [reply noxfile.py]: yes now I use only `git ls-file`

**Comment by DavideTisi** [comment]
I simplified the code take a look @SanggyuChong @PicoCentauri @Luthaf

File: noxfile.py [hunk 2]
```diff
@@ -171,6 +205,9 @@ def lint(session):
         session.install("isort")
         session.install("sphinx-lint")
 
+    # Get files
+    LINT_FILES = get_lint_files() + get_list_of_files()
```
Comment (Luthaf) [inline noxfile.py]: I would change `get_lint_files` to return both

File: noxfile.py [hunk 3]
```diff
@@ -19,23 +20,56 @@
...
+        "ipynb-to-gallery.py",
+        "generate-gallery.py",
+        "noxfile.py",
+        "docs/src/conf.py",
+        "developer",
+    ]
+    return LINT_FILES
+
+
+def filter_files(tracked_files):
```
Comment (Luthaf) [inline noxfile.py:42]: I still don't understand why we have this function?

File: noxfile.py [hunk 4]
```diff
@@ -19,23 +20,56 @@
...
+                and file.split("/")[-1] != "README.rst"
+            ):
+                returns.append(file)
+
+    return returns
+
+
+# We want to mimic
+# git ls-files examples
+def get_list_of_files():
```
Comment (Luthaf) [inline noxfile.py]:
```suggestion
def get_example_files():
```

File: noxfile.py (hunk 3 above)
Comment (DavideTisi) [reply noxfile.py:42]: there are files with are not ignored by git that I wanted to ignore by the linter. 
For ex the traj files or the yml files, also `git ls-files` contains also all the `.gitignore` files that i do not want to lint

Comment (Luthaf) [reply noxfile.py:42]: ah, because `git ls-files` gives you files which are not Python?

Comment (Luthaf) [reply noxfile.py:42]: Then I'm fine with this, maybe add some comment/docstring explaining the function?

Comment (DavideTisi) [reply noxfile.py:42]: yes gives you everything is tracked by `git`
```
>>> git ls-files
.github/workflows/comment-docs-download.yml
.github/workflows/docs.yml
.github/workflows/lint.yml
.gitignore
CONDUCT.md
CONTRIBUTING.rst
LICENSE
README.rst
developer/get_examples.py
docs/requirements.txt
docs/src/.gitignore
docs/src/conf.py
docs/src/index.rst.in
examples/batch-cp2k/.gitignore
examples/batch-cp2k/README.rst
examples/batch-cp2k/environment.yml
examples/batch-cp2k/example.xyz
examples/batch-cp2k/reference-trajectory.py
examples/batch-cp2k/reftraj_template.cp2k
examples/batch-cp2k/run_calcs.sh
examples/dos-align/README.rst
examples/dos-align/dos-align.py
examples/dos-align/environment.yml
examples/gaas-map/.gitignore
examples/gaas-map/README.rst
examples/gaas-map/environment.yml
examples/gaas-map/gaas-map.py
examples/lode-linear/README.rst
examples/lode-linear/charge-charge.xyz
examples/lode-linear/environment.yml
examples/lode-linear/lode-linear.py
examples/lpr/.gitignore
examples/lpr/README.rst
examples/lpr/environment.yml
examples/lpr/lpr.py
examples/roy-gch/.gitignore
examples/roy-gch/README.rst
examples/roy-gch/environment.yml
examples/roy-gch/roy-gch.py
examples/sample-selection/.gitignore
examples/sample-selection/README.rst
examples/sample-selection/environment.yml
examples/sample-selection/input-fps.xyz
examples/sample-selection/sample-selection.py
generate-gallery.py
ipynb-to-gallery.py
noxfile.py
pyproject.toml
```

**Review by Luthaf — APPROVED** [review]

**Pull request merged**